components/resource_control/src/lib.rs

```diff
@@ -71,58 +69,82 @@ impl ResourceController {
-            self.resource_groups.remove(name).map(|kv| kv.1)
+            self.resource_groups.remove(name).map(|(_, v)| v)
         }
     }
 
+    pub fn total_cpu_quota(&self) -> f64 {
+        self.total_cpu_quota
+    }
+
+    pub fn get_resource_group(&self, name: &str) -> Option<Ref<String, ResourceGroupConfig>> {
+        self.resource_groups.get(name)
+    }
+
     pub fn get_all_resource_groups(&self) -> Vec<ResourceGroupConfig> {
-        self.resource_groups
-            .iter()
-            .map(|g| g.config.clone())
-            .collect()
+        self.resource_groups.iter().map(|g| g.clone()).collect()
+    }
+}
+
+pub struct ResourceController {
+    manager: Arc<ResourceGroupManager>,
+    // record consumption of each resource group
+    resource_consumptions: DashMap<String, ResourceGroup>,
+    last_min_vt: AtomicU64,
+    start_ts: Instant,
+    // the value is the duration delta(in ms) from start_ts
+    last_vt_update_time: AtomicU64,
+}
+
+impl ResourceController {
+    pub fn new(manager: Arc<ResourceGroupManager>) -> Self {
+        Self {
+            manager,
+            resource_consumptions: DashMap::new(),
+            last_min_vt: AtomicU64::new(0),
+            start_ts: Instant::now_coarse(),
+            last_vt_update_time: AtomicU64::new(0),
+        }
     }
 
     #[inline]
-    fn resource_group(&self, name: &str) -> Ref<String, Arc<ResourceGroup>> {
-        self.resource_groups
-            .get(name)
-            .unwrap_or_else(|| self.resource_groups.get("default").unwrap())
-
-        // if let Some(group) = self.resource_groups.get(&group_id) {
-        //     return group;
-        // }
-
-        // self.add_resource_group(ResourceGroupConfig::new(
-        //     group_id,
-        //     "".into(),
-        //     self.total_cpu_quota,
-        //     0,
-        //     0,
-        // ));
-        // self.resource_groups.get(&group_id).unwrap()
+    fn resource_group(&self, name: &str) -> Ref<String, ResourceGroup> {
+        if let Some(g) = self.resource_consumptions.get(name) {
+            return g;
+        } else {
+            // get the resource group config from the manager
+            if let Some(g) = self.manager.get_resource_group(name) {
+                let config = g.value();
+                let priority_factor =
+                    (self.manager.total_cpu_quota() / config.cpu_quota * 100.0) as u64;
+                let group = ResourceGroup {
+                    priority_factor,
+                    virtual_time: AtomicU64::new(self.last_min_vt.load(Ordering::Acquire)),
+                };
+                self.resource_consumptions
+                    .insert(config.name.to_lowercase(), group);
+                return self.resource_consumptions.get(name).unwrap();
+            } else {
+                self.resource_consumptions.get("default").unwrap()
+            }
+        }
     }
 
     pub fn get_priority(&self, name: &str, priority: CommandPri) -> u64 {
```

Comment:
The previous will add a constant priority delta to the group virtual time at each `get_priority` call, the default value represent 100ms. Do the default const value also suitable for the write phase?

Reply:
Yes, need to test to see how it's performed. Plan to give the difference priority delta to different resource types after making resource controller generic